### PR TITLE
Fb1263

### PR DESCRIPTION
**Pull request by spireken**
I'm not sure why MatrixTransformTester.cpp is showing up as a new file.

### CODE REVIEW

File: Include/Ashkal/Shapes/Circle.hpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#ifndef ASHKAL_CIRCLE_HPP
+#define ASHKAL_CIRCLE_HPP
+#include "Ashkal/Shapes/Shape.hpp"
+
+namespace Ashkal {
+
+  //! Represents a Circle.
```
Comment (kranar) [inline Include/Ashkal/Shapes/Circle.hpp]: It's noteworthy to specify that it's a unit circle centered at the origin.

File: Include/Ashkal/Shapes/Line.hpp [hunk 2]
```diff
@@ -5,28 +5,19 @@
 
 namespace Ashkal {
 
-  //! Represents a Line on a 2D plane.
+  //! Represents a line.
```
Comment (kranar) [inline Include/Ashkal/Shapes/Line.hpp]: See Circle comment.

File: Include/Ashkal/Shapes/Square.hpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#ifndef ASHKAL_SQUARE_HPP
+#define ASHKAL_RECTANGLE_HPP
+#include "Ashkal/Shapes/Shape.hpp"
+
+namespace Ashkal {
+
+  //! Represents a Square.
```
Comment (kranar) [inline Include/Ashkal/Shapes/Square.hpp]: See Circle comment.

File: Source/Shapes/Line.cpp [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
 void Line::accept(ShapeVisitor& visitor) const {
   visitor.visit(*this);
 }
+
+Line Ashkal::make_line(const Point& p1, const Point& p2) {
+  auto l = Line();
+  auto length = std::sqrt(std::pow(std::abs(p1.x - p2.x), 2) +
+    std::pow(std::abs(p1.y - p2.y), 2));
+  auto slope = (p1.y - p2.y) / (p1.x - p2.x);
+  auto angle = (std::atan(1) * 4) - std::abs(atan(slope) - atan(1));
```
Comment (kranar) [inline Source/Shapes/Line.cpp]: ```scalar * f(x)``` never ```f(x) * scalar```

File: Source/Shapes/Line.cpp [hunk 5]
```diff
@@ -1,18 +1,17 @@
...
-
 void Line::accept(ShapeVisitor& visitor) const {
   visitor.visit(*this);
 }
+
+Line Ashkal::make_line(const Point& p1, const Point& p2) {
+  auto l = Line();
+  auto length = std::sqrt(std::pow(std::abs(p1.x - p2.x), 2) +
+    std::pow(std::abs(p1.y - p2.y), 2));
+  auto slope = (p1.y - p2.y) / (p1.x - p2.x);
```
Comment (kranar) [inline Source/Shapes/Line.cpp:13]: This won't work for a vertical line. Write a unit test to exercise a vertical line case and then fix the implementation.

File: Source/Tests/MakeShapeTester.cpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  auto p2 = Point{2, -2};
+  auto line = make_line(p1, p2);
+  auto slope = (p1.y - p2.y) / (p1.x - p2.x);
+  auto angle = (std::atan(1) * 4) - std::abs(atan(slope) - atan(1));
+  auto result = Matrix(
+    {std::cos(angle) * std::sqrt(32), std::sqrt(32), -2},
+    {-std::sqrt(32), std::cos(angle) * std::sqrt(32), 2},
+    {0, 0, 1});
+  REQUIRE(equals(line.get_transformation(), result));
+}
```
Comment (kranar) [inline Source/Tests/MakeShapeTester.cpp:38]: This test assumes that your implementation is correct which is not a valid way to go about testing things. It's very challenging to come up with a proper test case that is not biased by your implementation but we have to keep the two of them distinct.

In this case, what can we say should be true about our transformation? Well if transformation T takes a line L1 = (-0.5, -0.5) -> (0.5, 0.5) and transforms it into line L2 = p1 -> p2, then T^-1 should take the point p1 and transform it into (-0.5, -0.5) and take the point p2 and transform it into point (0.5, 0.5).

Note that this fact has nothing to do with our implementation and doesn't require us to come up with magic numbers or calculate any kind of cosines or sqrts. It just creates two points, p1, p2... and checks that the inverse transformation takes p1 to (-0.5, -0.5) and takes p2 to (0.5, 0.5).

**Review by kranar — CHANGES_REQUESTED** [review]
Good good stuff.

File: Source/Tests/MakeShapeTester.cpp (hunk 6 above)
Comment (spireken) [reply Source/Tests/MakeShapeTester.cpp:38]: So the inverse transformation of the created line should be equal the transformation of the default line?

**Comment by kranar** [comment]
Resubmit when all the changes have been made.

**Pull request closed without merging**